@Aeirx — welcome, and sorry for the dead air. Your CI was not slow, it was sitting behind the first-time-contributor approval gate: all ten runs sat at `action_required` and nothing had started. Approved, they are running now. That was on me, not on anything you did.

Answering your two open questions first so you are not blocked on either while I read the gate itself.

**RUF036 at `delegation_scope.py:425`** — already fixed and in flight on #3994 (`int | None | str` → `int | str | None`; ruff 0.15.8 started flagging it last week). Leave it alone, you would only collide.

**`token_type` at `agent_identity.py:329`** — reproduced on clean `main`:

```
src/bernstein/core/agents/agent_identity.py:329: error: Argument "token_type" to
"AgentCredential" has incompatible type "str"; expected "Literal['opaque', 'jwt']"
```

You were right not to fold it into a security change, and it is not only a typing nit. `from_dict` is where a record on disk becomes an object the rest of the code trusts, and every field beside it is coerced *and* constrained (`_credential_tenant_id`, `_string_list`) — `token_type` is only coerced. A record carrying any string at all loads clean and then reaches `if cred.token_type != "jwt"` at `:825`, where an unrecognised value takes a branch instead of being refused.

Filed as #4004 with the enum-refusal decision left open, and it is yours if you want it — say the word and I will assign you. (Nothing is red about it today because `uv run mypy src` is advisory in `ci.yml` and the strict zone covers `core/identity/`, not `core/agents/`.)

**On the Windows-only failures** — running the verify scope against stashed `main` for a like-for-like comparison is the right way to have shown that, and 19-on-both-sides is a real answer rather than an assurance. The Linux runners will settle it shortly; if anything survives there I will chase it rather than hand it back.

I have assigned you #3914 so nobody duplicates the work.

Full review of the gate follows once I have read it properly. Starting from agreement on the important part: containment at merge rather than prevention at write is exactly what #3781 settled on, and "the out-of-scope write still happens on disk, it just does not reach the repository" is the sentence that section of the operations guide has been missing.

--- a/docs/operations/security-and-identity.md
+++ b/docs/operations/security-and-identity.md
@@ -460,6 +460,17 @@ never become a signed scope. A pattern already on disk that cannot be
 interpreted matches nothing, and the refusal names it — an unreadable scope
 must not widen into "no scope".
 
+**A damaged record refuses; a missing one does not.** The two are not the same
+evidence. No record means no scope was ever declared for that session. A record
+that is on disk and does not load — truncated by an interrupted write, hand-
+edited so its two copies of the scope disagree, or sitting in a directory the
+orchestrator cannot list — is a scope someone declared that the gate cannot
+read, and it refuses rather than merging unbounded. The refusal is journalled
+under `allowed-files-unreadable` instead of `allowed-files-scope`, so
+`.sdd/runtime/refused_merges.jsonl` distinguishes "the agent went outside its
+scope" from "the scope itself needs repair". Repair it with the identity
+commands above, or revoke and re-spawn the agent.
+
 ## Delegation capability tokens
 
 When a run fans out, authority fans out with it. A **capability token** makes

--- a/src/bernstein/core/agents/spawner_merge.py
+++ b/src/bernstein/core/agents/spawner_merge.py
@@ -103,11 +103,21 @@ def _incoming_change(worktree_root: Path, branch: str) -> tuple[list[str], str]:
     the score describes this merge rather than the branch's whole history.
     An unreadable branch yields an empty change, which the scorer treats as
     a zero-risk one; the gate below refuses only on evidence.
+
+    ``--no-renames`` on the file list, because rename detection reports only
+    a rename's *destination*. Every gate reading this list judges paths, and
+    a list that omits the path a merge removes lets ``git mv outside inside``
+    pass a check that ``git rm outside`` fails -- the disguised removal
+    admitted and the honest one refused. A rename is two paths changing, so
+    both are named. The diff body keeps rename detection: it is scored as
+    text rather than judged as paths, and inflating a pure rename into a
+    delete-plus-add there would change what a blast radius means without
+    closing anything.
     """
     from bernstein.core.git_ops import run_git
 
     spec = f"HEAD...{branch}"
-    names = run_git(["diff", "--name-only", spec], worktree_root, timeout=30)
+    names = run_git(["diff", "--name-only", "--no-renames", spec], worktree_root, timeout=30)
     if names.returncode != 0:
         return [], ""
     files = [line.strip() for line in names.stdout.splitlines() if line.strip()]
@@ -173,7 +183,13 @@ def _signed_file_scope(worktree_root: Path, session_id: str) -> list[str] | None
 
     try:
         identity = AgentIdentityStore(auth_dir).get(session_id)
-    except OSError as exc:  # an unreadable store is not a scope
+    except OSError as exc:
+        # An unreadable store is not a scope. Failing open is safe here only
+        # because the store sits on the orchestrator's side of the boundary:
+        # ``worktree_root`` is the merge target, not the agent's worktree, so
+        # the branch being merged cannot reach ``.sdd/auth`` to provoke this.
+        # A gate that read its policy from ground the attacker can write would
+        # need the opposite default.
         logger.debug("Could not read agent identities for %s: %s", _sanitise_for_log(session_id), exc)
         return None
     return None if identity is None else identity.allowed_files

--- a/src/bernstein/core/agents/spawner_merge.py
+++ b/src/bernstein/core/agents/spawner_merge.py
@@ -11,7 +11,7 @@ import logging
 import threading
 import time
 from pathlib import Path
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Final
 
 from bernstein.core.git_ops import MergeResult, merge_with_conflict_detection
 from bernstein.core.models import AgentBackend, AgentSession
@@ -163,13 +163,33 @@ def _blast_radius_refusal(
 # ---------------------------------------------------------------------------
 
 
-def _signed_file_scope(worktree_root: Path, session_id: str) -> list[str] | None:
+class _UnreadableScope:
+    """A scope that exists for the session but cannot be read back.
+
+    Distinct from ``None``, which is "no identity record was ever written".
+    Both leave the gate with no patterns to match against, and both have to
+    land on opposite sides of it: an absent record is the settled open
+    default, while a record someone wrote and the store cannot parse is the
+    same failure :func:`~bernstein.core.path_scope.paths_outside_scope`
+    already answers by admitting nothing. An unreadable scope must not widen
+    into "no scope", whether what is unreadable is one pattern or the whole
+    record carrying it.
+    """
+
+    __slots__ = ()
+
+
+_UNREADABLE_SCOPE: Final = _UnreadableScope()
+
+
+def _signed_file_scope(worktree_root: Path, session_id: str) -> list[str] | _UnreadableScope | None:
     """Return the ``allowed_files`` scope signed for ``session_id``.
 
     ``None`` means no scope was ever declared, which is not the same as an
     empty one: an empty list is a credential that deliberately restricts
     nothing, while ``None`` is the absence of a credential at all. Both are
     unrestricted here, but only the first is a statement someone made.
+    :data:`_UNREADABLE_SCOPE` is neither, and refuses.
 
     The store is opened only when its directory already exists. Constructing
     an :class:`AgentIdentityStore` mints a JWT secret as a side effect, and a
@@ -184,15 +204,52 @@ def _signed_file_scope(worktree_root: Path, session_id: str) -> list[str] | None
     try:
         identity = AgentIdentityStore(auth_dir).get(session_id)
     except OSError as exc:
-        # An unreadable store is not a scope. Failing open is safe here only
-        # because the store sits on the orchestrator's side of the boundary:
-        # ``worktree_root`` is the merge target, not the agent's worktree, so
-        # the branch being merged cannot reach ``.sdd/auth`` to provoke this.
-        # A gate that read its policy from ground the attacker can write would
-        # need the opposite default.
-        logger.debug("Could not read agent identities for %s: %s", _sanitise_for_log(session_id), exc)
-        return None
-    return None if identity is None else identity.allowed_files
+        logger.error("Could not read agent identities for %s: %s", _sanitise_for_log(session_id), exc)
+        return _UNREADABLE_SCOPE
+    if identity is not None:
+        return identity.allowed_files
+
+    # ``get`` answers ``None`` for a session that never had an identity and
+    # for one whose record is on disk but does not deserialise -- a file
+    # truncated by a crash mid-write, or one whose two copies of the scope
+    # disagree, is skipped by the store's shared reader exactly like a record
+    # that is not there. An unreadable directory likewise makes every record
+    # look absent. Only genuine absence is the settled open default, so the
+    # directory is read rather than the difference guessed at.
+    try:
+        present = any(entry.name == f"{session_id}.json" for entry in (auth_dir / "agent_identities").iterdir())
+    except OSError as exc:
+        logger.error("Could not list agent identities for %s: %s", _sanitise_for_log(session_id), exc)
+        return _UNREADABLE_SCOPE
+    return _UNREADABLE_SCOPE if present else None
+
+
+def _refuse_merge(
+    session: AgentSession,
+    worktree_root: Path,
+    branch: str,
+    *,
+    reason: str,
+    code: str,
+) -> MergeResult:
+    """Record, log, and score one file-scope refusal.
+
+    Shared by the gate's two refusals so a scope that is out of bounds and a
+    scope that cannot be read leave the same evidence behind: an operator
+    reconstructing either from ``refused_merges.jsonl`` should not find one
+    of them better recorded than the other.
+    """
+    _record_merge_refusal(worktree_root, session.id, branch, reason=code)
+    logger.error(
+        "Refusing to merge agent work from %s: %s",
+        _sanitise_for_log(session.id),
+        _sanitise_for_log(reason),
+    )
+    from bernstein.core.metric_collector import get_collector
+
+    for task_id in session.task_ids:
+        get_collector().record_merge_result(task_id, success=False)
+    return MergeResult(success=False, conflicting_files=[], error=reason)
 
 
 def _file_scope_refusal(
@@ -207,6 +264,10 @@ def _file_scope_refusal(
     scope is empty. Every identity minted before this gate existed carries an
     empty scope, so the gate is inert until an operator sets one.
 
+    A record that is on disk and does not resolve is neither of those, and
+    refuses: it is a scope someone declared, and the gate cannot see what it
+    said.
+
     The refusal is containment rather than prevention. The out-of-scope write
     already happened inside the agent's own worktree; what this stops is that
     change reaching the repository, and the branch is left intact so an
@@ -215,6 +276,17 @@ def _file_scope_refusal(
     from bernstein.core.path_scope import paths_outside_scope
 
     patterns = _signed_file_scope(worktree_root, session.id)
+    if isinstance(patterns, _UnreadableScope):
+        return _refuse_merge(
+            session,
+            worktree_root,
+            branch,
+            reason=(
+                f"refused: identity {session.id!r} has a signed file scope on disk that "
+                f"cannot be read back; an unreadable scope is not an absent one"
+            ),
+            code="allowed-files-unreadable",
+        )
     if not patterns:
         return None
 
@@ -223,21 +295,16 @@ def _file_scope_refusal(
     if not outside:
         return None
 
-    reason = (
-        f"refused: identity {session.id!r} is scoped to {', '.join(patterns)}, "
-        f"but the merge would bring in {', '.join(outside)}"
+    return _refuse_merge(
+        session,
+        worktree_root,
+        branch,
+        reason=(
+            f"refused: identity {session.id!r} is scoped to {', '.join(patterns)}, "
+            f"but the merge would bring in {', '.join(outside)}"
+        ),
+        code="allowed-files-scope",
     )
-    _record_merge_refusal(worktree_root, session.id, branch, reason="allowed-files-scope")
-    logger.error(
-        "Refusing to merge agent work from %s: %s",
-        _sanitise_for_log(session.id),
-        _sanitise_for_log(reason),
-    )
-    from bernstein.core.metric_collector import get_collector
-
-    for task_id in session.task_ids:
-        get_collector().record_merge_result(task_id, success=False)
-    return MergeResult(success=False, conflicting_files=[], error=reason)
 
 
 # ---------------------------------------------------------------------------

--- a/tests/unit/security/test_merge_file_scope_gate.py
+++ b/tests/unit/security/test_merge_file_scope_gate.py
@@ -161,6 +161,47 @@ def test_a_pattern_stored_before_validation_existed_matches_nothing_rather_than_
     assert "src/ok.py" in result.error
 
 
+def test_a_rename_out_of_scope_is_refused_like_a_deletion(tmp_path: Path) -> None:
+    """Moving a file into scope is still removing it from where it was.
+
+    Rename detection reports only a rename's destination, so a scope check
+    reading that list would see an in-scope path and admit a merge that
+    deletes an out-of-scope one — passing the disguised removal while
+    refusing the honest ``git rm``. Both orderings must land the same way.
+    """
+    _repo_with_agent_branch(tmp_path, "s10", {"infra/deploy.tf": "terraform\n"})
+    _mint(tmp_path, "s10", ["src/**"])
+
+    # Baseline: deleting it outright is already refused.
+    deletion = _refuse(tmp_path, "s10")
+    assert deletion is not None
+    assert "infra/deploy.tf" in deletion.error
+
+    # The same removal, disguised as a move into scope.
+    moved = tmp_path / "moved"
+    moved.mkdir()
+    _run(["git", "init", "-b", "main"], moved)
+    _run(["git", "config", "user.email", "test@example.com"], moved)
+    _run(["git", "config", "user.name", "Test User"], moved)
+    (moved / "infra").mkdir()
+    (moved / "infra" / "deploy.tf").write_text("terraform\n", encoding="utf-8")
+    _run(["git", "add", "-A"], moved)
+    _run(["git", "commit", "-m", "init"], moved)
+
+    _run(["git", "checkout", "-b", "agent/s11"], moved)
+    (moved / "src").mkdir()
+    _run(["git", "mv", "infra/deploy.tf", "src/deploy.tf"], moved)
+    _run(["git", "commit", "-m", "rename"], moved)
+    _run(["git", "checkout", "main"], moved)
+
+    _mint(moved, "s11", ["src/**"])
+
+    result = _refuse(moved, "s11")
+
+    assert result is not None, "a rename out of scope must not pass a check a deletion fails"
+    assert "infra/deploy.tf" in result.error
+
+
 def test_a_refused_merge_leaves_the_branch_intact(tmp_path: Path) -> None:
     """Containment, not destruction: an operator has to be able to look at
     what was refused."""

--- a/tests/unit/security/test_merge_file_scope_gate.py
+++ b/tests/unit/security/test_merge_file_scope_gate.py
@@ -234,3 +234,93 @@ def test_the_scope_check_runs_against_the_same_file_list_the_refusal_reports(tmp
     assert outside, "the fixture must produce at least one out-of-scope path"
     for path in outside:
         assert path in result.error
+
+
+def _truncate_record(record: Path) -> None:
+    """A write that did not finish - a crash or a full disk mid-``_save``."""
+    record.write_text('{"id": "s12", "role": "backend"', encoding="utf-8")
+
+
+def _disagree_with_credential(record: Path) -> None:
+    """One field edited, so the record now carries two answers to "scoped to what"."""
+    stored = json.loads(record.read_text(encoding="utf-8"))
+    stored["allowed_files"] = []
+    record.write_text(json.dumps(stored), encoding="utf-8")
+
+
+def _unparseable_enum(record: Path) -> None:
+    """A value the record's own reader refuses, in a field that is not the scope."""
+    stored = json.loads(record.read_text(encoding="utf-8"))
+    stored["status"] = "not-a-real-status"
+    record.write_text(json.dumps(stored), encoding="utf-8")
+
+
+@pytest.mark.parametrize(
+    "corrupt",
+    [_truncate_record, _disagree_with_credential, _unparseable_enum],
+    ids=["truncated", "scope-disagrees-with-credential", "unparseable-enum"],
+)
+def test_an_identity_record_that_does_not_deserialise_refuses_rather_than_widening(
+    tmp_path: Path,
+    corrupt: Any,
+) -> None:
+    """The fail-open direction again, one level up from the pattern.
+
+    The store's shared reader answers ``None`` for a record it cannot parse
+    and for a record that is not there, and only the second is the settled
+    open default. A scope switched off by a truncated write - or by a
+    one-field edit - would leave an operator believing an agent is bounded
+    while every path merges. Refusing is recoverable; widening is silent.
+    """
+    _repo_with_agent_branch(tmp_path, "s12", {"infra/deploy.tf": "y\n"})
+    _mint(tmp_path, "s12", ["src/**"])
+    corrupt(tmp_path / ".sdd" / "auth" / "agent_identities" / "s12.json")
+
+    result = _refuse(tmp_path, "s12")
+
+    assert result is not None, "an unreadable record must not widen to 'no record'"
+    assert result.success is False
+    assert "s12" in result.error
+
+
+def test_an_unreadable_identity_directory_refuses_rather_than_widening(tmp_path: Path) -> None:
+    """A directory the gate cannot list makes every record look absent.
+
+    ``get`` reports no identity either way, so a gate that trusted it would
+    treat a store it cannot read as a store with nothing in it - the widest
+    possible reading of the least trustworthy evidence.
+    """
+    _repo_with_agent_branch(tmp_path, "s13", {"infra/deploy.tf": "y\n"})
+    _mint(tmp_path, "s13", ["src/**"])
+    identities = tmp_path / ".sdd" / "auth" / "agent_identities"
+
+    identities.chmod(0o000)
+    try:
+        if any(entry.name for entry in identities.iterdir()):  # pragma: no cover - root or a permissive fs
+            pytest.skip("filesystem does not enforce directory permissions for this user")
+    except OSError:
+        pass
+    try:
+        result = _refuse(tmp_path, "s13")
+    finally:
+        identities.chmod(0o755)
+
+    assert result is not None, "an unlistable store must not widen to 'no scope'"
+
+
+def test_an_unreadable_scope_is_recorded_under_its_own_reason(tmp_path: Path) -> None:
+    """Two different refusals, two different reasons in the refusal journal.
+
+    An operator reading ``refused_merges.jsonl`` has to be able to tell "the
+    agent went outside its scope" from "the scope itself is damaged": the
+    first is the agent's problem and the second is the operator's.
+    """
+    _repo_with_agent_branch(tmp_path, "s14", {"infra/deploy.tf": "y\n"})
+    _mint(tmp_path, "s14", ["src/**"])
+    _truncate_record(tmp_path / ".sdd" / "auth" / "agent_identities" / "s14.json")
+
+    assert _refuse(tmp_path, "s14") is not None
+
+    journal = tmp_path / ".sdd" / "runtime" / "refused_merges.jsonl"
+    reasons = [json.loads(line)["reason"] for line in journal.read_text(encoding="utf-8").splitlines() if line.strip()]
+    assert reasons == ["allowed-files-unreadable"]

